File: Data Structures & Algorithms/coin-change-ii/submission-4.py
class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        n = len(coins)
        tab = [[0] * (amount + 1) for _ in range(n + 1)]
        if amount == 0:
            return 1

        for i in range(n - 1, -1, -1):
            for curr in range(amount - 1, -1, -1):
                if curr + coins[i] == amount:
                    tab[i][curr] = 1 + tab[i + 1][curr]
                else:
                    tab[i][curr] = (tab[i][curr + coins[i]] if curr + coins[i] <= amount else 0) + tab[i + 1][curr]

        return tab[0][0]

File: Data Structures & Algorithms/target-sum/submission-4.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int n = nums.length;
        int tot = Arrays.stream(nums).sum();

        if (Math.abs(target) > tot)
        {
            return 0;
        }

        int[][] tab = new int[nums.length + 1][tot * 2 + 1];
        tab[n][target + tot] = 1;

        for (int i = n - 1; i >= 0; i--)
        {
            for (int curr = -tot; curr < tot + 1; curr++)
            {
                int add = curr + nums[i];
                int sub = curr - nums[i];

                int future_add = (-tot <= add && add <= tot) ? tab[i + 1][add + tot] : 0;
                int future_sub = (-tot <= sub && sub <= tot) ? tab[i + 1][sub + tot] : 0;

                tab[i][curr + tot] = future_add + future_sub;
            }
        }

        return tab[0][0 + tot];
    }
}
